File: mobile/lib/features/transactions/data/models/transaction.dart
import '../../../../core/utils/money.dart';
import 'transaction_status.dart';
import 'transaction_type.dart';

class Transaction {
  const Transaction({
    required this.id,
    required this.contactId,
    required this.contactName,
    required this.type,
    required this.originalAmount,
    required this.remainingAmount,
    required this.currency,
    required this.description,
    required this.transactionDate,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      contactId: json['contactId'] as String,
      contactName: json['contactName'] as String,
      type: TransactionType.fromApiValue(json['type'] as String),
      originalAmount: parseAmount(json['originalAmount']),
      remainingAmount: parseAmount(json['remainingAmount']),
      currency: json['currency'] as String,
      description: json['description'] as String?,
      transactionDate: DateTime.parse(json['transactionDate'] as String),
      status: TransactionStatus.fromApiValue(json['status'] as String),
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  final String id;
  final String contactId;
  final String contactName;
  final TransactionType type;
  final num originalAmount;
  final num remainingAmount;
  final String currency;
  final String? description;
  final DateTime transactionDate;
  final TransactionStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;

  bool get isPaid => status == TransactionStatus.paid;
}

File: backend/src/main/java/com/lendly/transaction/domain/Transaction.java
package com.lendly.transaction.domain;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.lendly.contact.domain.Contact;
import com.lendly.identity.domain.User;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "transactions")
@Getter
@Setter
@NoArgsConstructor
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(optional = false)
    @JoinColumn(name = "contact_id", nullable = false)
    private Contact contact;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 20)
    private TransactionType type;

    @Column(name = "original_amount", nullable = false, precision = 14, scale = 2)
    private BigDecimal originalAmount;

    @Column(name = "remaining_amount", nullable = false, precision = 14, scale = 2)
    private BigDecimal remainingAmount;

    @Column(name = "currency", nullable = false, length = 3)
    private String currency = "TND";

    @Column(name = "description", length = 500)
    private String description;

    @Column(name = "transaction_date", nullable = false)
    private LocalDate transactionDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private TransactionStatus status;

    @Version
    @Column(name = "version", nullable = false)
    private long version;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    public Transaction(
        User user,
        Contact contact,
        TransactionType type,
        BigDecimal originalAmount,
        String currency,
        String description,
        LocalDate transactionDate
    ) {
        this.user = user;
        this.contact = contact;
        this.type = type;
        this.originalAmount = originalAmount;
        this.remainingAmount = originalAmount;
        this.currency = currency;
        this.description = description;
        this.transactionDate = transactionDate;
        this.status = TransactionStatus.ACTIVE;
    }
}
